fastlane/lib/fastlane/plugins/plugin_manager.rb

```diff
@@ -394,5 +401,19 @@ def store_plugin_reference(gem_name)
         actions: references
       }
     end
+
+    private
+
+    # recursively collect runtime dependencies
+    def collect_bundler_runtime_dependencies(collection, source)
+      runtime_deps = source.dependencies.select { |d| d.type == :runtime }
+      collection.concat(runtime_deps)
+      runtime_deps.each do |d|
```

Comment:
it's possible to return to the previous non recursive behavior by disabling this loop. In case we want to ship this optionally.

Reply:
There _shouldn't_ be a case where we run into an infinite loop here, right? 🤔 

Could we add an if statement here to not run `collect_bundler_runtime_dependencies` again if `d` is already in the collection?

Reply:
I made a quick test, with 2 projects dependending on each other. Code and execution below:

```
$ find .. -type f -exec sh -c 'for i; do echo "\n=== $i\n"; cat $i; done' sh {} +

=== ../1/1.gemspec

# frozen_string_literal: true
Gem::Specification.new do |spec|
  spec.name        = 'a1'
  spec.version     = '1.0.0'
  spec.authors     = ["x"]
  spec.email       = 'x@example.com'

  spec.required_ruby_version = '>= 2.5.0'

  spec.summary     = "1"
  #spec.description = U3d::DESCRIPTION

  spec.add_dependency 'a2'
end

=== ../1/test.rb

require 'bundler'

def gemfile_path
  # This is pretty important, since we don't know what kind of
  # Gemfile the user has (e.g. Gemfile, gems.rb, or custom env variable)
  Bundler::SharedHelpers.default_gemfile.to_s
rescue Bundler::GemfileNotFound
  nil
end

def lockfile_path
  Bundler::SharedHelpers.default_lockfile.to_s
rescue Bundler::GemfileNotFound
  nil
end

def collect_bundler_runtime_dependencies(collection, source)
  runtime_deps = source.dependencies.select { |d| d.type == :runtime }
  collection.concat(runtime_deps)
  runtime_deps.each do |d|
    collect_bundler_runtime_dependencies(collection, d.to_spec)
  rescue Gem::MissingSpecError
  # ignoring unresolvable dependencies
  end
  collection
end

puts gemfile_path
puts lockfile_path
dsl = Bundler::Dsl.evaluate(gemfile_path, lockfile_path, true)
dependencies = collect_bundler_runtime_dependencies([], dsl).uniq
puts dependencies.map(&:name).join(",")

=== ../1/Gemfile

source "https://rubygems.org"


gemspec
gem "a2", path: "../2"

=== ../2/2.gemspec

# frozen_string_literal: true
Gem::Specification.new do |spec|
  spec.name        = 'a2'
  spec.version     = '1.0.0'
  spec.authors     = ["x"]
  spec.email       = 'x@example.com'

  spec.required_ruby_version = '>= 2.5.0'

  spec.summary     = "2"
  #spec.description = U3d::DESCRIPTION

  spec.add_dependency 'a1'
end

=== ../2/Gemfile

source "https://rubygems.org"

gemspec
gem "a1", path: "../1"
```


execution

```
lacostej@MacBook-Pro-10 ~/Code/test/1 (dev2021) bundle install
Resolving dependencies...
Your bundle requires gems that depend on each other, creating an infinite loop. Please remove either gem 'a1' or gem 'a2' and try again.
lacostej@MacBook-Pro-10 ~/Code/test/1 (dev2021) ruby test.rb 
/Users/lacostej/Code/test/1/Gemfile
/Users/lacostej/Code/test/1/Gemfile.lock
a1,a2

lacostej@MacBook-Pro-10 ~/Code//test/1 (dev2021) cd ../2/
lacostej@MacBook-Pro-10 ~/Code/test/2 (dev2021) ruby ../1/test.rb 
/Users/lacostej/Code/test/2/Gemfile
/Users/lacostej/Code/test/2/Gemfile.lock
a2,a1
lacostej@MacBook-Pro-10 ~/Code/test/2 (dev2021) bundle install
Resolving dependencies...
Your bundle requires gems that depend on each other, creating an infinite loop. Please remove either gem 'a1' or gem 'a2' and try again.
```

so bundle install doesn't work, no Gemfile.lock gets generated. And the code doesn't loop.

Reply:
I added the code to prevent the potential recursive call, in case.

Reply:
For information the code was 
```
      # prevent potential infinite recursion. Should never happen. Better safe than sorry.
      runtime_deps -= collection
```